site/content/blog/2019-10-21-localizing-matterpoll/index.md

```diff
@@ -0,0 +1,188 @@
+
+```go
+// getUserLocalizer returns a localizer that localizes in the users locale
+func (p *MatterpollPlugin) getUserLocalizer(userID string) *i18n.Localizer {
+	user, err := p.API.GetUser(userID)
+	if err != nil {
+		p.API.LogWarn("Failed get user's locale", "error", err.Error())
+		return p.getServerLocalizer()
+	}
+
+	return i18n.NewLocalizer(p.bundle, user.Locale)
+}
+```
+
+Not every part of Matterpoll is localizable on a user level. Take polls for example.
+
+{{< figure src="poll-1.png" alt="Poll 1">}}
+
+While the response when clicking an option should be localized with the user's local, the "Total Votes" text cannot. Posts are shown in the same way for every user, hence we have to use the same localization for every user. Admins can configure a [Default Client Language](https://docs.mattermost.com/administration/config-settings.html#default-client-language) in the System Console that is used for newly-created users and pages where the user hasn’t logged in. We decided to use this local for translating these kind of strings and build a helper for this:
+
+```go
+// getServerLocalizer returns a localizer that localizes in the server default client locale
+func (p *MatterpollPlugin) getServerLocalizer() *i18n.Localizer {
+	return i18n.NewLocalizer(p.bundle, *p.ServerConfig.LocalizationSettings.DefaultClientLocale)
+}
+```
+
+With this preparation we were finally ready to do the actual task: Defining and translating actual strings. Putting the pieces together is quite simple and comes down to:
+
+```go
+	l := p.getUserLocalizer(userID)
+	response, err := l.LocalizeMessage(&i18n.Message{
+		ID:    "response.vote.counted",
+		Other: "Your vote has been counted.",
+	}
+	
+	// Send response back
+```
+
+Most but not all translation strings can be translated without any context information: for example, the "Total Votes" text you saw above include the number of votes. We did incorporate this information directly into the `Message`:
+
+```go
+	l := p.getUserLocalizer(userID)
+	response, err := l.Localize(&i18n.LocalizeConfig{
+		DefaultMessage: &i18n.Message{
+			ID:    "poll.message.totalVotes",
+			Other: "**Total votes**: {{.TotalVotes}}",
+		},
+		TemplateData:   map[string]interface{}{"TotalVotes": numberOfVotes},
+	})
+		
+	// Send response back
+```
+
+### Extracting the Translation Strings
+
+As already mentioned [`goi18n`](https://github.com/nicksnyder/go-i18n#command-goi18n-), allow the extraction of the translation string from the code into the translation file. This can be done by running
```

Comment:
```suggestion
As already mentioned [`goi18n`](https://github.com/nicksnyder/go-i18n#command-goi18n-), allows the extraction of the translation string from the code into the translation file. This can be done by running:
```